code/modules/reagents/chemistry/holder.dm

```diff
@@ -385,181 +689,510 @@
 	var/list/cached_reagents = reagent_list
 	for(var/reagent in cached_reagents)
 		var/datum/reagent/R = reagent
-		R.on_move (A, Running)
+		R.on_move(A, Running)
 	update_total()
 
 /datum/reagents/proc/conditional_update(atom/A)
 	var/list/cached_reagents = reagent_list
 	for(var/reagent in cached_reagents)
 		var/datum/reagent/R = reagent
-		R.on_update (A)
+		R.on_update(A)
 	update_total()
 
+/// Handle any reactions possible in this holder
+/// Also UPDATES the reaction list
+/// High potential for infinite loopsa if you're editing this.
 /datum/reagents/proc/handle_reactions()
+	if(QDELING(src))
+		CRASH("[my_atom] is trying to handle reactions while being flagged for deletion. It presently has [length(reagent_list)] number of reactants in it. If that is over 0 then something terrible happened.")
+
+	if(!length(reagent_list))//The liver is calling this method a lot, and is often empty of reagents so it's pointless busywork. It should be an easy fix, but I'm nervous about touching things beyond scope. Also since everything is so handle_reactions() trigger happy it might be a good idea having this check anyways.
+		return FALSE
+
 	if(flags & NO_REACT)
-		return //Yup, no reactions here. No siree.
+		if(is_reacting)
+			force_stop_reacting() //Force anything that is trying to to stop
+		return FALSE //Yup, no reactions here. No siree.
+
+	if(is_reacting)//Prevent wasteful calculations
+		if(datum_flags != DF_ISPROCESSING)//If we're reacting - but not processing (i.e. we've transfered)
+			START_PROCESSING(SSreagents, src)
+		if(!(has_changed_state()))
+			return FALSE
 
 	var/list/cached_reagents = reagent_list
 	var/list/cached_reactions = GLOB.chemical_reactions_list
 	var/datum/cached_my_atom = my_atom
+	LAZYNULL(failed_but_capable_reactions)
 
-	var/reaction_occurred = 0
-	do
-		var/list/possible_reactions = list()
-		reaction_occurred = 0
-		for(var/reagent in cached_reagents)
-			var/datum/reagent/R = reagent
-			for(var/reaction in cached_reactions[R.type]) // Was a big list but now it should be smaller since we filtered it with our reagent id
-				if(!reaction)
-					continue
+	. = 0
+	var/list/possible_reactions = list()
+	for(var/_reagent in cached_reagents)
+		var/datum/reagent/reagent = _reagent
+		for(var/_reaction in cached_reactions[reagent.type]) // Was a big list but now it should be smaller since we filtered it with our reagent id
+			if(!_reaction)
+				continue
```

Comment:
suggested change continues into next comment because git still doesn't let you select across removal diffs 😢 
```suggestion
		for(var/datum/chemical_reaction/reaction as() in cached_reactions[reagent.type]) // Was a big list but now it should be smaller since we filtered it with our reagent id
```